### DISCUSSION TITLE
Creating segmented upload directories by hash

### DISCUSSION THREAD

**Discussion by dschissler** [pocketbase/pocketbase#1404] [Q&A]
Typically a file system directory will begin slowing down once many thousands of files are in it.  This can be minimized by creating subdirectories that use the first 1-3 hash/code characters.  So a id of `abcdef` would be in the `ab` directory.  This can allow it to scale better.

There might be newer filesystems that index better and so alleviate this issue, but I'm not aware of it.

**Comment by ganigeorgiev** [comment]
> Typically a file system directory will begin slowing down once many thousands of files are in it.

I'm not able to find information on this or a benchmark that shows significant impact on the performance in case of longer directory/file names.

**Reply by dschissler** [reply]
I recall that it happened when there were tens of thousands of directories in one directory.  That was a thing on ext4, but I don't know if btrfs, NTFS and others have this limitation.

**Reply by ganigeorgiev** [reply]
@dschissler I cannot find any info on this and I don't think it will have any significant impact on the performance if we reduce the directory/file names and/or combine directories based on some hash.

I'm ok to restructure how we store the files but there need to be a convincing reason for that because it'll add a fair amount of complexity to the files management.

Currently we store everything under `:collectionId/:recordId` and this is easy to understand and manage. In case of collection/record delete we can delete easily all stored files by just deleting their related directories/prefixes.

**Reply by dschissler** [reply]
@ganigeorgiev Sorry to bother you on some obsolete issues.  The best that I could find was this:

> It makes no difference until you hit the ext3 32,000 names per directory limit. Upgrading to ext4 can get around that, as well as the other benefits ext4 has.

So basically modern filesystems don't have this limitation, or it's at least been pushed up significantly.

**Reply by dschissler** [reply]
Unfortunately this reveals just how old that I've become.
